src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -1,34 +1,208 @@
+		    _multiDocumentParser = multiDocumentParser;
 		    _database = database;
-	    }
+		    _stream = stream;
+		    _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+			
+			_log = _database.LoggerSetup.GetLogger<IncomingReplicationHandler>(_database.Name);
+			_cts = CancellationTokenSource.CreateLinkedTokenSource(_database.DatabaseShutdown);
+		}
 
 	    public void Start()
 	    {
-		    
+		    _incomingThread = new Thread(ReceiveReplicatedDocuments)
+		    {
+				IsBackground = true,
+				Name = $"Incoming replication {FromToString}"
+		    };
+			_incomingThread.Start();
+	    }
+
+		//TODO : do not forget to add logging and code to record stats
+	    private void ReceiveReplicatedDocuments()
+	    {
+			using (_context)
+			using (var writer = new BlittableJsonTextWriter(_context, _stream))
+			using (_multiDocumentParser)
+			{
+				try
+				{
+					while (!_cts.IsCancellationRequested)
+					{
+						using (var message = _multiDocumentParser.ParseToMemory("IncomingReplication/read-message"))
+						{
+							_cts.Token.ThrowIfCancellationRequested();
+							string messageType;
+							if (!message.TryGet(Constants.MessageType, out messageType))
+								throw new InvalidOperationException(
+									$"Received replication message without type. All messages must have property {Constants.MessageType}");
+
+							switch (messageType)
+							{
+								case "GetLastEtag":
+									_context.Write(_writer, new DynamicJsonValue
+									{
+										[Constants.Replication.PropertyNames.LastSentEtag] = GetLastReceivedEtag(Guid.Parse(_incomingMessageHeader.SourceDatabaseId), _context),
+										[Constants.MessageType] = "GetLastEtag"
+									});
+									_writer.Flush();
+									break;
+								case "ReplicationBatch":
+
+									BlittableJsonReaderArray replicatedDocs;
+									if (!message.TryGet(Constants.Replication.PropertyNames.ReplicationBatch, out replicatedDocs))
+										throw new InvalidDataException(
+											$"Expected the message to have a field with replicated document array, named {Constants.Replication.PropertyNames.ReplicationBatch}. The property wasn't found");
+
+									using (_context.OpenWriteTransaction())
+									{
+										ReceiveDocuments(_context, replicatedDocs);
+										_context.Transaction.Commit();
+									}
+
+									break;
+								default:
+									throw new InvalidOperationException("Unrecognized replication message type.");
+							}
+						}
+					}
+				}
+				catch (Exception e)
+				{
+					OnFailed(e, this);
+				}
+				finally
+				{
+					_context = null;
+					_stream = null;
+				}
+			}
+	    }
+
+	    public string FromToString => $"from {_connectionInfo.SourceDatabaseName} at {_connectionInfo.SourceUrl} (into database {_database.Name})";
+
+	    private long GetLastReceivedEtag(Guid srcDbId, DocumentsOperationContext context)
+	    {
+		    var dbChangeVector = _database.DocumentsStorage.GetDatabaseChangeVector(context);
```

Comment:
Might be better to have a method that check for this directly, instead of loading everything, then searching in memory